# misclassified-variants
Introduction: Databases used for clinical interpretation in oncology care primarily derive genetic data from patients of European ancestry, leading to biases in genetic research and clinical practice. One practical issue that subsequently arises is the misclassification of multi-ancestral population variants as tumor-associated because they are not represented in standard reference genomes.
Methods: To systematically find misclassified variants, we merged cancer-related somatic variants from the Catalogue of Somatic Mutations in Cancer (COSMIC) with multi-ancestral population variants from the Genome Aggregation Databasesâ€™ Linkage Disequilibrium (GnomAD). By comparing genomic coordinates, reference and alternate alleles, we were able to confirm and report misclassified variants. 
Results: From the 1,906,732 variants in COSMIC, we identified 6,957 variants that aligned with GnomAD's normal population variants. Notably, African ancestral populations in GnomAD had 17,478,395 variants, with 5,320 (0.03%) variants misclassified, while East Asian populations had 9,239,894 variants, with 4132 (0.04%) variants misclassified. An analysis revealed that the African ancestral populations had the highest number of misclassified variants at 76.47%, while the East Asian population had the least at 59.39%. 
Conclusion: This initial step in the comparison of variants from COSMIC to the GnomAD population databases, will support a more accurate interpretation of genomic sequencing data and reduce the bias related to genomic ancestry.
